internal/build/setup.go

```diff
@@ -0,0 +1,299 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package build
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"strings"
+	"time"
+
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/yourbase/commons/xcontext"
+	"github.com/yourbase/narwhal"
+	"github.com/yourbase/yb/internal/buildcontext"
+	"github.com/yourbase/yb/internal/ybtrace"
+	"github.com/yourbase/yb/plumbing"
+	"go.opentelemetry.io/otel/api/trace"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/label"
+	"zombiezen.com/go/log"
+)
+
+// ContextCloser is a buildcontext.Context that has resources that need to be
+// cleaned up.
+type ContextCloser interface {
+	buildcontext.Context
+	io.Closer
+}
+
+// PhaseDeps defines the dependencies for a build target phase.
+//
+// TODO(ch2285): This should be moved out of this package and into a separate
+// loader package.
+type PhaseDeps struct {
+	TargetName          string
+	Resources           map[string]*narwhal.ContainerDefinition
+	EnvironmentTemplate map[string]string
+	// TODO(ch2744): Buildpacks
+}
+
+// Setup arranges for the phase's dependencies to be available, returning a new
+// build context that has the dependencies configured. It is the caller's
+// responsibility to call Close on the returned build context.
+func Setup(ctx context.Context, g G, phase *PhaseDeps) (_ ContextCloser, err error) {
+	ctx, span := ybtrace.Start(ctx, "Setup "+phase.TargetName, trace.WithAttributes(
+		label.String("target", phase.TargetName),
+	))
+	defer func() {
+		if err != nil {
+			span.SetStatus(codes.Unknown, err.Error())
+		}
+		span.End()
+	}()
+
+	expContainers, closeFunc, err := startContainers(ctx, g.DockerClient, g.Stderr, g.DockerNetworkID, phase.Resources)
+	if err != nil {
+		return nil, fmt.Errorf("setup %s: %w", phase.TargetName, err)
+	}
+	defer func() {
+		if err != nil && closeFunc != nil {
+			closeFunc()
+		}
+	}()
+	exp := configExpansion{
+		Containers: expContainers,
+	}
+	env := buildcontext.Environment{
+		Vars: make(map[string]string),
+	}
+	for k, t := range phase.EnvironmentTemplate {
+		v, err := exp.expand(t)
+		if err != nil {
+			return nil, fmt.Errorf("setup %s: expand %s: %w", phase.TargetName, k, err)
+		}
+		env.Vars[k] = v
+	}
+	return contextCloser{
+		Context: buildcontext.EnvContext{
+			Context: g.Context,
+			Env:     env,
+		},
+		closeFunc: closeFunc,
+	}, nil
+}
+
+type container struct {
+	client       *docker.Client
+	resourceName string
+	id           string
+	ip           net.IP
+}
+
+func startContainers(ctx context.Context, client *docker.Client, output io.Writer, networkID string, defs map[string]*narwhal.ContainerDefinition) (_ containersExpansion, _ func() error, err error) {
```

Comment:
Added more comments, and plumbed a `G` because the arguments weren't adding clarity. I can't think of a way to split the function in a way that adds clarity, so leaving as-is.